### ISSUE TITLE
Infinite partition synchronization between nodes

### ISSUE THREAD

**Issue by skywalkery** [hazelcast/hazelcast#4426]
Hi guys,
I use hazelcast version 3.3.3 and have a problem with infinite partition synchronization between two nodes. I don't know how to reproduce this bug, sometime it works, sometime not.
I start one node, then I start second node, and have infinite process. By the way, I use MapStore which stored data into amazon s3.

Log from the first node:

```
2015-01-16 11:57:29,695|INFO |main|c.h.i.DefaultAddressPicker|[LOCAL] [hzGroup] [3.3.3] Prefer IPv4 stack is true.
2015-01-16 11:57:29,705|INFO |main|c.h.i.DefaultAddressPicker|[LOCAL] [hzGroup] [3.3.3] Picked Address[192.168.34.12]:5701, using socket ServerSocket[addr=/0.0.0.0,localport=5701], bind any local is true
2015-01-16 11:57:29,745|INFO |main|c.h.s.i.BasicOperationScheduler|[192.168.34.12]:5701 [hzGroup] [3.3.3] Starting with 2 generic operation threads and 2 partition operation threads.
2015-01-16 11:57:29,833|INFO |main|com.hazelcast.system|[192.168.34.12]:5701 [hzGroup] [3.3.3] Hazelcast 3.3.3 (20141112 - eadb69c) starting at Address[192.168.34.12]:5701
2015-01-16 11:57:29,833|INFO |main|com.hazelcast.system|[192.168.34.12]:5701 [hzGroup] [3.3.3] Copyright (C) 2008-2014 Hazelcast.com
2015-01-16 11:57:29,850|INFO |main|c.hazelcast.instance.Node|[192.168.34.12]:5701 [hzGroup] [3.3.3] Creating MulticastJoiner
2015-01-16 11:57:29,855|INFO |main|c.h.core.LifecycleService|[192.168.34.12]:5701 [hzGroup] [3.3.3] Address[192.168.34.12]:5701 is STARTING
2015-01-16 11:57:33,307|INFO |main|c.h.c.MulticastJoiner|[192.168.34.12]:5701 [hzGroup] [3.3.3]


Members [1] {
        Member [192.168.34.12]:5701 this
}

2015-01-16 11:57:33,350|INFO |main|c.h.core.LifecycleService|[192.168.34.12]:5701 [hzGroup] [3.3.3] Address[192.168.34.12]:5701 is STARTED
2015-01-16 11:57:33,420|INFO |main|c.h.p.InternalPartitionService|[192.168.34.12]:5701 [hzGroup] [3.3.3] Initializing cluster partition table first arrangement...
2015-01-16 11:59:18,887|INFO |hz.container.javaee.util.HazelcastUtil.IO.thread-Acceptor|c.h.n.tcp.SocketAcceptor|[192.168.34.12]:5701 [hzGroup] [3.3.3] Accepting socket connection from /192.168.34.13:41688
2015-01-16 11:59:18,893|INFO |hz.container.javaee.util.HazelcastUtil.cached.thread-29|c.h.n.t.TcpIpConnectionManager|[192.168.34.12]:5701 [hzGroup] [3.3.3] Established socket connection between /192.168.34.12:5
701 and /192.168.34.13:41688
2015-01-16 11:59:25,914|INFO |hz.container.javaee.util.HazelcastUtil.generic-operation.thread-1|c.h.c.ClusterService|[192.168.34.12]:5701 [hzGroup] [3.3.3]

Members [2] {
        Member [192.168.34.12]:5701 this
        Member [192.168.34.13]:5701
}

2015-01-16 11:59:26,915|WARN |AsyncAppender-Worker-Thread-2|c.h.c.ClusterService|[192.168.34.12]:5701 [hzGroup] [3.3.3] While waiting finalize join calls...   at com.hazelcast.spi.impl.BasicInvocationFuture.resolveApplicationRes
ponse(BasicInvocationFuture.java:321)
2015-01-16 11:59:30,024|INFO |hz.container.javaee.util.HazelcastUtil.cached.thread-29|c.h.p.InternalPartitionService|[192.168.34.12]:5701 [hzGroup] [3.3.3] Remaining migration tasks in queue => 135
2015-01-16 11:59:30,174|INFO |hz.container.javaee.util.HazelcastUtil.migration|c.h.p.InternalPartitionService|[192.168.34.12]:5701 [hzGroup] [3.3.3] Re-partitioning cluster data... Migration queue size: 135
2015-01-16 11:59:45,024|INFO |hz.container.javaee.util.HazelcastUtil.cached.thread-30|c.h.p.InternalPartitionService|[192.168.34.12]:5701 [hzGroup] [3.3.3] Remaining migration tasks in queue => 134
//infinite part
2015-01-16 12:00:15,024|INFO |hz.container.javaee.util.HazelcastUtil.cached.thread-8|c.h.p.InternalPartitionService|[192.168.34.12]:5701 [hzGroup] [3.3.3] Remaining migration tasks in queue => 134
//
2015-01-16 12:00:45,024|INFO |hz.container.javaee.util.HazelcastUtil.cached.thread-16|c.h.p.InternalPartitionService|[192.168.34.12]:5701 [hzGroup] [3.3.3] Remaining migration tasks in queue => 134
2015-01-16 12:00:49,015|WARN |hz.container.javaee.util.HazelcastUtil.partition-operation.thread-0|c.h.s.i.BasicInvocation|[192.168.34.12]:5701 [hzGroup] [3.3.3] Retrying invocation: BasicInvocation{ serviceName
='hz:impl:mapService', op=com.hazelcast.map.operation.MapSizeOperation@6908e7c9, partitionId=0, replicaIndex=0, tryCount=250, tryPauseMillis=500, invokeCount=100, callTimeout=60000, target=Address[192.168.34.12]:5701}, Reason: com.hazelc
ast.spi.exception.PartitionMigratingException: Partition is migrating! this:Address[192.168.34.12]:5701, partitionId: 0, operation: com.hazelcast.map.operation.MapSizeOperation, service: hz:impl:mapService
2015-01-16 12:01:30,190|WARN |hz.container.javaee.util.HazelcastUtil.cached.thread-19|c.h.s.i.BasicInvocation|[192.168.34.12]:5701 [hzGroup] [3.3.3] No response for 120000 ms. BasicInvocationFuture{invocation=B
asicInvocation{ serviceName='hz:core:partitionService', op=com.hazelcast.partition.impl.MigrationOperation{partitionId=0, migration=MigrationInfo{ partitionId=0, source=Address[192.168.34.12]:5701, destination=Address[192.168.34.13]:5701
, master=Address[192.168.34.12]:5701, valid=true, processing=false}, compressed=true}, partitionId=0, replicaIndex=0, tryCount=250, tryPauseMillis=1000, invokeCount=1, callTimeout=60000, target=Address[192.168.34.13]:5701}, done=false}
2015-01-16 12:01:30,190|WARN |hz.container.javaee.util.HazelcastUtil.cached.thread-19|c.h.s.i.BasicInvocation|[192.168.34.12]:5701 [hzGroup] [3.3.3] Asking if operation execution has been started: BasicInvocati
onFuture{invocation=BasicInvocation{ serviceName='hz:core:partitionService', op=com.hazelcast.partition.impl.MigrationOperation{partitionId=0, migration=MigrationInfo{ partitionId=0, source=Address[192.168.34.12]:5701, destination=Addres
s[192.168.34.13]:5701, master=Address[192.168.34.12]:5701, valid=true, processing=false}, compressed=true}, partitionId=0, replicaIndex=0, tryCount=250, tryPauseMillis=1000, invokeCount=1, callTimeout=60000, target=Address[192.168.34.13]
:5701}, done=false}
2015-01-16 12:01:31,965|WARN |http-8080-5|c.h.s.i.BasicInvocation|[192.168.34.12]:5701 [hzGroup] [3.3.3] 'is-executing': false -> BasicInvocationFuture{invocation=BasicInvocation{ serviceName='hz:impl:mapService', op=com.hazelca
st.map.operation.QueryOperation@3f274683, partitionId=-1, replicaIndex=0, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeout=60000, target=Address[192.168.34.13]:5701}, done=false}
```

Log from the second node:

```
2015-01-16 11:59:18,622|INFO |main|com.hazelcast.system|[192.168.34.13]:5701 [hzGroup] [3.3.3] Hazelcast 3.3.3 (20141112 - eadb69c) starting at Address[192.168.34.13]:5701
2015-01-16 11:59:18,622|INFO |main|com.hazelcast.system|[192.168.34.13]:5701 [hzGroup] [3.3.3] Copyright (C) 2008-2014 Hazelcast.com
2015-01-16 11:59:18,635|INFO |main|c.hazelcast.instance.Node|[192.168.34.13]:5701 [hzGroup] [3.3.3] Creating MulticastJoiner
2015-01-16 11:59:18,644|INFO |main|c.h.core.LifecycleService|[192.168.34.13]:5701 [hzGroup] [3.3.3] Address[192.168.34.13]:5701 is STARTING
2015-01-16 11:59:18,893|INFO |main|c.h.c.MulticastJoiner|[192.168.34.13]:5701 [hzGroup] [3.3.3] Trying to join to discovered node: Address[192.168.34.12]:5701
2015-01-16 11:59:18,908|INFO |hz.javaee.util.HazelcastUtil.cached.thread-2|c.h.n.tcp.SocketConnector|[192.168.34.13]:5701 [hzGroup] [3.3.3] Connecting to /192.168.34.12:5701, timeout: 0, b
ind-any: true
2015-01-16 11:59:18,916|INFO |hz.javaee.util.HazelcastUtil.cached.thread-2|c.h.n.t.TcpIpConnectionManager|[192.168.34.13]:5701 [hzGroup] [3.3.3] Established socket connection between /192.
168.34.13:41688 and /192.168.34.12:5701
2015-01-16 11:59:25,937|INFO |hz.javaee.util.HazelcastUtil.generic-operation.thread-0|c.h.c.ClusterService|[192.168.34.13]:5701 [hzGroup] [3.3.3]

Members [2] {
        Member [192.168.34.12]:5701
        Member [192.168.34.13]:5701 this
}

2015-01-16 11:59:28,118|INFO |main|c.h.core.LifecycleService|[192.168.34.13]:5701 [hzGroup] [3.3.3] Address[192.168.34.13]:5701 is STARTED
2015-01-16 11:59:30,064|WARN |hz.javaee.util.HazelcastUtil.partition-operation.thread-0|c.h.p.InternalPartitionService|[192.168.34.13]:5701 [hzGroup] [3.3.3] Sync replica target is null, n
o need to sync -> partition: 176, replica: 1
2015-01-16 11:59:30,064|WARN |hz.javaee.util.HazelcastUtil.partition-operation.thread-0|c.h.p.InternalPartitionService|[192.168.34.13]:5701 [hzGroup] [3.3.3] Sync replica target is null, n
o need to sync -> partition: 178, replica: 1
2015-01-16 11:59:30,067|WARN |hz.javaee.util.HazelcastUtil.partition-operation.thread-0|c.h.p.InternalPartitionService|[192.168.34.13]:5701 [hzGroup] [3.3.3] Sync replica target is null, n
o need to sync -> partition: 216, replica: 1
2015-01-16 11:59:30,068|WARN |hz.javaee.util.HazelcastUtil.partition-operation.thread-0|c.h.p.InternalPartitionService|[192.168.34.13]:5701 [hzGroup] [3.3.3] Sync replica target is null, n
o need to sync -> partition: 224, replica: 1
2015-01-16 11:59:30,068|WARN |hz.javaee.util.HazelcastUtil.partition-operation.thread-0|c.h.p.InternalPartitionService|[192.168.34.13]:5701 [hzGroup] [3.3.3] Sync replica target is null, n
o need to sync -> partition: 230, replica: 1
2015-01-16 11:59:30,068|WARN |hz.javaee.util.HazelcastUtil.partition-operation.thread-0|c.h.p.InternalPartitionService|[192.168.34.13]:5701 [hzGroup] [3.3.3] Sync replica target is null, n
o need to sync -> partition: 236, replica: 1
2015-01-16 11:59:30,068|WARN |hz.javaee.util.HazelcastUtil.partition-operation.thread-0|c.h.p.InternalPartitionService|[192.168.34.13]:5701 [hzGroup] [3.3.3] Sync replica target is null, n
o need to sync -> partition: 262, replica: 1
2015-01-16 11:59:30,070|WARN |hz.javaee.util.HazelcastUtil.partition-operation.thread-1|c.h.p.InternalPartitionService|[192.168.34.13]:5701 [hzGroup] [3.3.3] Sync replica target is null, n
o need to sync -> partition: 137, replica: 1
2015-01-16 11:59:30,073|WARN |hz.javaee.util.HazelcastUtil.partition-operation.thread-1|c.h.p.InternalPartitionService|[192.168.34.13]:5701 [hzGroup] [3.3.3] Sync replica target is null, n
o need to sync -> partition: 139, replica: 1
2015-01-16 11:59:30,077|WARN |hz.javaee.util.HazelcastUtil.partition-operation.thread-1|c.h.p.InternalPartitionService|[192.168.34.13]:5701 [hzGroup] [3.3.3] Sync replica target is null, n
o need to sync -> partition: 143, replica: 1
2015-01-16 11:59:30,081|WARN |hz.javaee.util.HazelcastUtil.partition-operation.thread-1|c.h.p.InternalPartitionService|[192.168.34.13]:5701 [hzGroup] [3.3.3] Sync replica target is null, n
o need to sync -> partition: 145, replica: 1
2015-01-16 11:59:30,083|WARN |hz.javaee.util.HazelcastUtil.partition-operation.thread-1|c.h.p.InternalPartitionService|[192.168.34.13]:5701 [hzGroup] [3.3.3] Sync replica target is null, n
o need to sync -> partition: 155, replica: 1
2015-01-16 11:59:30,086|WARN |hz.javaee.util.HazelcastUtil.partition-operation.thread-1|c.h.p.InternalPartitionService|[192.168.34.13]:5701 [hzGroup] [3.3.3] Sync replica target is null, n
o need to sync -> partition: 183, replica: 1
2015-01-16 12:00:00,066|WARN |hz.javaee.util.HazelcastUtil.partition-operation.thread-1|c.h.p.InternalPartitionService|[192.168.34.13]:5701 [hzGroup] [3.3.3] Sync replica target is null, n
o need to sync -> partition: 137, replica: 1
2015-01-16 12:00:00,083|WARN |hz.javaee.util.HazelcastUtil.partition-operation.thread-1|c.h.p.InternalPartitionService|[192.168.34.13]:5701 [hzGroup] [3.3.3] Sync replica target is null, n
o need to sync -> partition: 139, replica: 1
...
infinite sync messages
```

Thread dump from the first node:

```
"akka.actor.default-dispatcher-8" prio=10 tid=0x00007eff94022800 nid=0x683d in Object.wait() [0x00007eff75787000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
        at java.lang.Object.wait(Native Method)
        - waiting on <0x000000077f1ec790> (a com.hazelcast.spi.impl.BasicInvocationFuture)
        at com.hazelcast.spi.impl.BasicInvocationFuture.pollResponse(BasicInvocationFuture.java:247)
        - locked <0x000000077f1ec790> (a com.hazelcast.spi.impl.BasicInvocationFuture)
        at com.hazelcast.spi.impl.BasicInvocationFuture.waitForResponse(BasicInvocationFuture.java:201)
        at com.hazelcast.spi.impl.BasicInvocationFuture.get(BasicInvocationFuture.java:180)
        at com.hazelcast.spi.impl.BasicInvocationFuture.get(BasicInvocationFuture.java:160)
        at com.hazelcast.map.BasicMapContextQuerySupport.getQueryResult(BasicMapContextQuerySupport.java:331)
        at com.hazelcast.map.BasicMapContextQuerySupport.addResultsOfPredicate(BasicMapContextQuerySupport.java:318)
        at com.hazelcast.map.BasicMapContextQuerySupport.query(BasicMapContextQuerySupport.java:229)
        at com.hazelcast.map.proxy.MapProxySupport.query(MapProxySupport.java:1057)
        at com.hazelcast.map.proxy.MapProxyImpl.keySet(MapProxyImpl.java:576)
        at com.hazelcast.map.proxy.MapProxyImpl.keySet(MapProxyImpl.java:558)
        at JobManagerImpl.getJobIds(JobManagerImpl.java:443)
        ...

   Locked ownable synchronizers:
        - None

"akka.actor.default-dispatcher-7" prio=10 tid=0x00007effa8f4d800 nid=0x683c in Object.wait() [0x00007eff75888000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
        at java.lang.Object.wait(Native Method)
        - waiting on <0x000000077fad9080> (a com.hazelcast.spi.impl.BasicInvocationFuture)
        at com.hazelcast.spi.impl.BasicInvocationFuture.pollResponse(BasicInvocationFuture.java:247)
        - locked <0x000000077fad9080> (a com.hazelcast.spi.impl.BasicInvocationFuture)
        at com.hazelcast.spi.impl.BasicInvocationFuture.waitForResponse(BasicInvocationFuture.java:201)
        at com.hazelcast.spi.impl.BasicInvocationFuture.get(BasicInvocationFuture.java:180)
        at com.hazelcast.spi.impl.BasicInvocationFuture.get(BasicInvocationFuture.java:160)
        at com.hazelcast.spi.impl.BasicOperationService$InvokeOnPartitions.retryFailedPartitions(BasicOperationService.java:522)
        at com.hazelcast.spi.impl.BasicOperationService$InvokeOnPartitions.invoke(BasicOperationService.java:460)
        at com.hazelcast.spi.impl.BasicOperationService$InvokeOnPartitions.access$600(BasicOperationService.java:430)
        at com.hazelcast.spi.impl.BasicOperationService.invokeOnAllPartitions(BasicOperationService.java:293)
        at com.hazelcast.map.proxy.MapProxySupport.size(MapProxySupport.java:616)
        at com.hazelcast.map.proxy.MapProxyImpl.size(MapProxyImpl.java:72)
        at JobManagerImpl$MapExtendedImpl.size(JobManagerImpl.java:536)

   Locked ownable synchronizers:
        - None

"pool-3-thread-1" prio=10 tid=0x00007eff80016000 nid=0x67c6 in Object.wait() [0x00007eff7ccfc000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
        at java.lang.Object.wait(Native Method)
        - waiting on <0x000000077c933410> (a com.hazelcast.spi.impl.BasicInvocationFuture)
        at com.hazelcast.spi.impl.BasicInvocationFuture.pollResponse(BasicInvocationFuture.java:247)
        - locked <0x000000077c933410> (a com.hazelcast.spi.impl.BasicInvocationFuture)
        at com.hazelcast.spi.impl.BasicInvocationFuture.waitForResponse(BasicInvocationFuture.java:201)
        at com.hazelcast.spi.impl.BasicInvocationFuture.get(BasicInvocationFuture.java:180)
        at com.hazelcast.spi.impl.BasicInvocationFuture.get(BasicInvocationFuture.java:160)
        at com.hazelcast.map.BasicMapContextQuerySupport.getQueryResult(BasicMapContextQuerySupport.java:331)
        at com.hazelcast.map.BasicMapContextQuerySupport.addResultsOfPredicate(BasicMapContextQuerySupport.java:318)
        at com.hazelcast.map.BasicMapContextQuerySupport.query(BasicMapContextQuerySupport.java:229)
        at com.hazelcast.map.proxy.MapProxySupport.query(MapProxySupport.java:1057)
        at com.hazelcast.map.proxy.MapProxyImpl.entrySet(MapProxyImpl.java:584)
        at com.hazelcast.map.proxy.MapProxyImpl.entrySet(MapProxyImpl.java:568)
        at java.util.concurrent.ConcurrentHashMap.putAll(ConcurrentHashMap.java:1161)
        at JobManagerImpl$2.run(JobManagerImpl.java:79)

   Locked ownable synchronizers:
        - <0x000000079ba849e8> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x000000079ba84f18> (a java.util.concurrent.ThreadPoolExecutor$Worker)

"hz.javaee.util.HazelcastUtil.migration" prio=10 tid=0x00007effa93bf000 nid=0x67af in Object.wait() [0x00007eff7e419000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
        at java.lang.Object.wait(Native Method)
        - waiting on <0x000000077464a2f0> (a com.hazelcast.spi.impl.BasicInvocationFuture)
        at com.hazelcast.spi.impl.BasicInvocationFuture.pollResponse(BasicInvocationFuture.java:247)
        - locked <0x000000077464a2f0> (a com.hazelcast.spi.impl.BasicInvocationFuture)
        at com.hazelcast.spi.impl.BasicInvocationFuture.waitForResponse(BasicInvocationFuture.java:201)
        at com.hazelcast.spi.impl.BasicInvocationFuture.get(BasicInvocationFuture.java:180)
        at com.hazelcast.partition.impl.InternalPartitionServiceImpl$MigrateTask.executeMigrateOperation(InternalPartitionServiceImpl.java:1631)
        at com.hazelcast.partition.impl.InternalPartitionServiceImpl$MigrateTask.run(InternalPartitionServiceImpl.java:1601)
        at com.hazelcast.partition.impl.InternalPartitionServiceImpl$MigrationThread.processTask(InternalPartitionServiceImpl.java:1741)
        at com.hazelcast.partition.impl.InternalPartitionServiceImpl$MigrationThread.doRun(InternalPartitionServiceImpl.java:1717)
        at com.hazelcast.partition.impl.InternalPartitionServiceImpl$MigrationThread.run(InternalPartitionServiceImpl.java:1696)

   Locked ownable synchronizers:
        - None

"hz.javaee.util.HazelcastUtil.wait-notify" prio=10 tid=0x00007effa934b000 nid=0x67ad waiting on condition [0x00007eff7e61b000]
   java.lang.Thread.State: TIMED_WAITING (parking)
        at sun.misc.Unsafe.park(Native Method)
        - parking to wait for  <0x000000079b8dd7c8> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
        at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:226)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2082)
        at java.util.concurrent.DelayQueue.poll(DelayQueue.java:256)
        at com.hazelcast.spi.impl.WaitNotifyServiceImpl$ExpirationTask.doRun(WaitNotifyServiceImpl.java:430)
        at com.hazelcast.spi.impl.WaitNotifyServiceImpl$ExpirationTask.run(WaitNotifyServiceImpl.java:415)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
        at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:334)
        at java.util.concurrent.FutureTask.run(FutureTask.java:166)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:724)
        at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
        at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:92)

   Locked ownable synchronizers:
        - <0x000000079b8de168> (a java.util.concurrent.ThreadPoolExecutor$Worker)
```

Thread dump from the second node:

```
"hz.javaee.util.HazelcastUtil.migration" prio=10 tid=0x00007f63893e5000 nid=0x38e9 waiting on condition [0x00007f635fefd000]
   java.lang.Thread.State: TIMED_WAITING (parking)
        at sun.misc.Unsafe.park(Native Method)
        - parking to wait for  <0x000000079c4980f8> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
        at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:226)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2082)
        at java.util.concurrent.LinkedBlockingQueue.poll(LinkedBlockingQueue.java:467)
        at com.hazelcast.partition.impl.InternalPartitionServiceImpl$MigrationThread.doRun(InternalPartitionServiceImpl.java:1712)
        at com.hazelcast.partition.impl.InternalPartitionServiceImpl$MigrationThread.run(InternalPartitionServiceImpl.java:1696)

   Locked ownable synchronizers:
        - None

"hz.javaee.util.HazelcastUtil.wait-notify" prio=10 tid=0x00007f6389377000 nid=0x38e7 waiting on condition [0x00007f637c199000]
   java.lang.Thread.State: TIMED_WAITING (parking)
        at sun.misc.Unsafe.park(Native Method)
        - parking to wait for  <0x000000079c44f700> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
        at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:226)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2082)
        at java.util.concurrent.DelayQueue.poll(DelayQueue.java:256)
        at com.hazelcast.spi.impl.WaitNotifyServiceImpl$ExpirationTask.doRun(WaitNotifyServiceImpl.java:430)
        at com.hazelcast.spi.impl.WaitNotifyServiceImpl$ExpirationTask.run(WaitNotifyServiceImpl.java:415)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
        at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:334)
        at java.util.concurrent.FutureTask.run(FutureTask.java:166)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:724)
        at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
        at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:92)

   Locked ownable synchronizers:
        - <0x000000079c451978> (a java.util.concurrent.ThreadPoolExecutor$Worker)

"hz.javaee.util.HazelcastUtil.partition-operation.thread-0" prio=10 tid=0x00007f638935a800 nid=0x38df waiting for monitor entry [0x00007f637c9a0000]
   java.lang.Thread.State: BLOCKED (on object monitor)
        at com.hazelcast.util.ConcurrencyUtil.getOrPutSynchronized(ConcurrencyUtil.java:38)
        - waiting to lock <0x000000079c52e8e8> (a java.util.concurrent.ConcurrentHashMap)
        at com.hazelcast.map.DefaultMapServiceContext.getMapContainer(DefaultMapServiceContext.java:71)
        at com.hazelcast.map.PartitionContainer$1.createNew(PartitionContainer.java:39)
        at com.hazelcast.map.PartitionContainer$1.createNew(PartitionContainer.java:37)
        at com.hazelcast.util.ConcurrencyUtil.getOrPutSynchronized(ConcurrencyUtil.java:40)
        - locked <0x000000079c625228> (a com.hazelcast.map.PartitionContainer)
        at com.hazelcast.map.PartitionContainer.getRecordStore(PartitionContainer.java:78)
        at com.hazelcast.map.DefaultMapServiceContext.getRecordStore(DefaultMapServiceContext.java:167)
        at com.hazelcast.map.operation.MapReplicationOperation.run(MapReplicationOperation.java:114)
        at com.hazelcast.partition.impl.MigrationOperation.runMigrationTask(MigrationOperation.java:191)
        at com.hazelcast.partition.impl.MigrationOperation.migrate(MigrationOperation.java:143)
        at com.hazelcast.partition.impl.MigrationOperation.doRun(MigrationOperation.java:93)
        at com.hazelcast.partition.impl.MigrationOperation.run(MigrationOperation.java:84)
        at com.hazelcast.spi.impl.BasicOperationService$OperationHandler.handle(BasicOperationService.java:699)
        at com.hazelcast.spi.impl.BasicOperationService$OperationHandler.access$400(BasicOperationService.java:675)
        at com.hazelcast.spi.impl.BasicOperationService$OperationPacketHandler.handle(BasicOperationService.java:649)
        at com.hazelcast.spi.impl.BasicOperationService$OperationPacketHandler.handle(BasicOperationService.java:604)
        at com.hazelcast.spi.impl.BasicOperationService$OperationPacketHandler.access$1400(BasicOperationService.java:591)
        at com.hazelcast.spi.impl.BasicOperationService$BasicDispatcherImpl.dispatch(BasicOperationService.java:543)
        at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.process(BasicOperationScheduler.java:439)
        at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.processPriorityMessages(BasicOperationScheduler.java:452)
        at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.doRun(BasicOperationScheduler.java:432)
        at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.run(BasicOperationScheduler.java:408)

   Locked ownable synchronizers:
        - None

"hz.javaee.util.HazelcastUtil.generic-operation.thread-1" prio=10 tid=0x00007f6389354000 nid=0x38de waiting for monitor entry [0x00007f637caa2000]
   java.lang.Thread.State: BLOCKED (on object monitor)
        at com.hazelcast.util.ConcurrencyUtil.getOrPutSynchronized(ConcurrencyUtil.java:38)
        - waiting to lock <0x000000079c52e8e8> (a java.util.concurrent.ConcurrentHashMap)
        at com.hazelcast.map.DefaultMapServiceContext.getMapContainer(DefaultMapServiceContext.java:71)
        at com.hazelcast.map.operation.AbstractMapOperation.beforeRun(AbstractMapOperation.java:39)
        at com.hazelcast.spi.impl.BasicOperationService$OperationHandler.beforeRun(BasicOperationService.java:711)
        at com.hazelcast.spi.impl.BasicOperationService$OperationHandler.handle(BasicOperationService.java:693)
        at com.hazelcast.spi.impl.BasicOperationService$OperationHandler.access$400(BasicOperationService.java:675)
        at com.hazelcast.spi.impl.BasicOperationService$OperationPacketHandler.handle(BasicOperationService.java:649)
        at com.hazelcast.spi.impl.BasicOperationService$OperationPacketHandler.handle(BasicOperationService.java:604)
        at com.hazelcast.spi.impl.BasicOperationService$OperationPacketHandler.access$1400(BasicOperationService.java:591)
        at com.hazelcast.spi.impl.BasicOperationService$BasicDispatcherImpl.dispatch(BasicOperationService.java:543)
        at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.process(BasicOperationScheduler.java:439)
        at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.doRun(BasicOperationScheduler.java:433)
        at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.run(BasicOperationScheduler.java:408)

   Locked ownable synchronizers:
        - None

**//infinite cycle here cause owner always null for some reason**
"hz.javaee.util.HazelcastUtil.generic-operation.thread-0" prio=10 tid=0x00007f6389352800 nid=0x38dd waiting on condition [0x00007f637cba2000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
        at java.lang.Thread.sleep(Native Method)
        at com.hazelcast.partition.impl.InternalPartitionServiceImpl.getPartitionOwnerOrWait(InternalPartitionServiceImpl.java:240)
        at com.hazelcast.map.DefaultMapServiceContext.isOwnedKey(DefaultMapServiceContext.java:200)
        at com.hazelcast.map.MapContainer.loadInitialKeys(MapContainer.java:214)
        at com.hazelcast.map.MapContainer.initMapStore(MapContainer.java:169)
        at com.hazelcast.map.MapContainer.initMapStoreOperations(MapContainer.java:114)
        at com.hazelcast.map.MapContainer.<init>(MapContainer.java:83)
        at com.hazelcast.map.DefaultMapServiceContext$1.createNew(DefaultMapServiceContext.java:33)
        at com.hazelcast.map.DefaultMapServiceContext$1.createNew(DefaultMapServiceContext.java:31)
        at com.hazelcast.util.ConcurrencyUtil.getOrPutSynchronized(ConcurrencyUtil.java:40)
        - locked <0x000000079c52e8e8> (a java.util.concurrent.ConcurrentHashMap)
        at com.hazelcast.map.DefaultMapServiceContext.getMapContainer(DefaultMapServiceContext.java:71)
        at com.hazelcast.map.proxy.MapProxySupport.<init>(MapProxySupport.java:140)
        at com.hazelcast.map.proxy.MapProxyImpl.<init>(MapProxyImpl.java:75)
        at com.hazelcast.map.MapRemoteService.createDistributedObject(MapRemoteService.java:26)
        at com.hazelcast.map.MapRemoteService.createDistributedObject(MapRemoteService.java:14)
        at com.hazelcast.map.MapService.createDistributedObject(MapService.java:126)
        at com.hazelcast.spi.impl.ProxyServiceImpl$ProxyRegistry.createProxy(ProxyServiceImpl.java:336)
        at com.hazelcast.spi.impl.ProxyServiceImpl$PostJoinProxyOperation.run(ProxyServiceImpl.java:574)
        at com.hazelcast.spi.impl.BasicOperationService$OperationHandler.handle(BasicOperationService.java:699)
        at com.hazelcast.spi.impl.BasicOperationService$OperationHandler.access$400(BasicOperationService.java:675)
        at com.hazelcast.spi.impl.BasicOperationService.runOperationOnCallingThread(BasicOperationService.java:229)
        at com.hazelcast.cluster.PostJoinOperation.run(PostJoinOperation.java:74)
        at com.hazelcast.spi.impl.BasicOperationService$OperationHandler.handle(BasicOperationService.java:699)
        at com.hazelcast.spi.impl.BasicOperationService$OperationHandler.access$400(BasicOperationService.java:675)
        at com.hazelcast.spi.impl.BasicOperationService.runOperationOnCallingThread(BasicOperationService.java:229)
        at com.hazelcast.cluster.FinalizeJoinOperation.run(FinalizeJoinOperation.java:89)
        at com.hazelcast.spi.impl.BasicOperationService$OperationHandler.handle(BasicOperationService.java:699)
        at com.hazelcast.spi.impl.BasicOperationService$OperationHandler.access$400(BasicOperationService.java:675)
        at com.hazelcast.spi.impl.BasicOperationService$OperationPacketHandler.handle(BasicOperationService.java:649)
        at com.hazelcast.spi.impl.BasicOperationService$OperationPacketHandler.handle(BasicOperationService.java:604)
        at com.hazelcast.spi.impl.BasicOperationService$OperationPacketHandler.access$1400(BasicOperationService.java:591)
        at com.hazelcast.spi.impl.BasicOperationService$BasicDispatcherImpl.dispatch(BasicOperationService.java:543)
        at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.process(BasicOperationScheduler.java:439)
        at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.processPriorityMessages(BasicOperationScheduler.java:452)
        at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.doRun(BasicOperationScheduler.java:432)
        at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.run(BasicOperationScheduler.java:408)

   Locked ownable synchronizers:
        - None

"main" prio=10 tid=0x00007f638800a800 nid=0x38c6 waiting for monitor entry [0x00007f6390023000]
   java.lang.Thread.State: BLOCKED (on object monitor)
        at com.hazelcast.util.ConcurrencyUtil.getOrPutSynchronized(ConcurrencyUtil.java:38)
        - waiting to lock <0x000000079c52e8e8> (a java.util.concurrent.ConcurrentHashMap)
        at com.hazelcast.map.DefaultMapServiceContext.getMapContainer(DefaultMapServiceContext.java:71)
        at com.hazelcast.map.proxy.MapProxySupport.<init>(MapProxySupport.java:140)
        at com.hazelcast.map.proxy.MapProxyImpl.<init>(MapProxyImpl.java:75)
        at com.hazelcast.map.MapRemoteService.createDistributedObject(MapRemoteService.java:26)
        at com.hazelcast.map.MapRemoteService.createDistributedObject(MapRemoteService.java:14)
        at com.hazelcast.map.MapService.createDistributedObject(MapService.java:126)
        at com.hazelcast.spi.impl.ProxyServiceImpl$ProxyRegistry.createProxy(ProxyServiceImpl.java:336)
        at com.hazelcast.spi.impl.ProxyServiceImpl$ProxyRegistry.getOrCreateProxy(ProxyServiceImpl.java:310)
        at com.hazelcast.spi.impl.ProxyServiceImpl.getDistributedObject(ProxyServiceImpl.java:131)
        at com.hazelcast.instance.HazelcastInstanceImpl.getDistributedObject(HazelcastInstanceImpl.java:388)
        at com.hazelcast.instance.HazelcastInstanceImpl.getMap(HazelcastInstanceImpl.java:176)
        at com.hazelcast.instance.HazelcastInstanceProxy.getMap(HazelcastInstanceProxy.java:71)
        at javaee.cache.HazelcastProvider.<init>(HazelcastProvider.java:28)

   Locked ownable synchronizers:
        - None
```

**Comment by skywalkery** [comment]
Some additional logs from the 2nd node:
```
2015-01-28 06:30:21,190|DEBUG|hz.container.javaee.util.HazelcastUtil.cached.thread-1|c.h.c.ClusterService|[192.168.34.13]:5701 [dev] [3.3.3] Sending MasterConfirmation to Member [192.168.34.12]
:5701
2015-01-28 06:30:22,908|DEBUG|hz.container.javaee.util.HazelcastUtil.partition-operation.thread-1|c.h.p.i.CheckReplicaVersion|[192.168.34.13]:5701 [dev] [3.3.3] Partition: 145 version is not ma
tching to version of the owner -> 0 -vs- 6
2015-01-28 06:30:22,911|WARN |hz.container.javaee.util.HazelcastUtil.partition-operation.thread-1|c.h.p.InternalPartitionService|[192.168.34.13]:5701 [dev] [3.3.3] Sync replica target is null,
no need to sync -> partition: 145, replica: 1
//...
//for the various partitions
//...
2015-01-28 06:30:44,432|DEBUG|hz.container.javaee.util.HazelcastUtil.partition-operation.thread-1|c.h.spi.impl.Backup|[192.168.34.13]:5701 [dev] [3.3.3] Wrong target! Backup{backupOp=null, orig
inalCaller=Address[192.168.34.12]:5701, version=[5, 0, 0, 0, 0, 0], sync=true} cannot be processed! Target should be: null
```

**Comment by Donnerbart** [comment]
Thanks for reporting your issue. So far I got the following information from you:
* Hazelcast version is 3.3.3
* Cluster size is 2
* Number of clients is 0
* Logs and stack traces are provided
* Steps to reproduce
  * Start first node
  * Start second node

You said you cannot reproduce this behavior all the time, but we would be grateful if you can provide a unit test and attach your configuration, so we may have at least a test which fails randomly.

Please also share with us the following information, to help us resolve your issue quickly and efficiently:
* Version of Java (also JVM parameters can be added)
* Operating System (for Linux kernel version will be helpful)
* Integration module versions if available (e.g. Tomcat, Jetty, Spring, Hibernate) with detailed configuration information (web.xml, Hibernate configuration, Spring context.xml etc.)

Could you please also try Hazelcast 3.4 and see if the issue is still present with that version?

**Comment by skywalkery** [comment]
Thanks for the response.

* java version "1.7.0_25"
Java(TM) SE Runtime Environment (build 1.7.0_25-b15)
Java HotSpot(TM) 64-Bit Server VM (build 23.25-b01, mixed mode)

* Linux 2.6.32-358.14.1.el6.x86_64 #1 SMP Tue Jul 16 23:51:20 UTC 2013 x86_64 x86_64 x86_64 GNU/Linux

* Tomcat 6.0

It looks like the issue does not reproduced on the version 3.4.
I am trying to create unit test for 3.3.3 which will be failed, but still without success.

**Comment by Donnerbart** [comment]
If you create a unit test within the Hazelcast repository you can use the annotation `@Repeat(100)` to run a test multiple times. I don't know by heart how to do this easily in JUnit. In TestNG you can use `@Test(invocationCount = 100)`. This often helps to track down non-deterministic failures.

Could you also provide some information how you integrated Hazelcast in Tomcat?

**Comment by ahmetmircik** [comment]
Hi,

Seems related to  https://github.com/hazelcast/hazelcast/issues/4348

Also keeping the info. below in mind may also be helpful:
__For the versions < 3.4:__
`InitialLoadMode#LAZY` --> blocks `getMap` till the end of the load of _all keys_,
`InitialLoadMode#EAGER` --> blocks `getMap` till the end of the load of _all key + value pairs_.

__For the versions >= 3.4__:
`#LAZY` --> does not block `getMap`,
`#EAGER` ---> blocks `getMap` till the end of the load of _all key + value pairs_ (same with < 3.4 behavior).

**Comment by skywalkery** [comment]
OK guys, I created the unit test that detect this problem. It happens when there is a delay for the node joining. I don't know how to get this delay in the legal way, but I saw it in the logs:
```
2015-01-29 06:05:14,257|DEBUG|hz.container.javaee.util.HazelcastUtil.generic-operation.thread-0|c.h.p.InternalPartitionService|[192.168.34.13]:5701 [dev] [3.3.3] Node should be active(true) and joined(false) to be able to process partition table!
```

So, I changed Node.setJoined() method and added 2s delay.
```java
package com.test;

import com.hazelcast.config.Config;
import com.hazelcast.config.MapConfig;
import com.hazelcast.config.MapStoreConfig;
import com.hazelcast.config.XmlConfigBuilder;
import com.hazelcast.core.Hazelcast;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.core.IMap;
import javassist.CannotCompileException;
import javassist.ClassPool;
import javassist.CtClass;
import javassist.CtMethod;
import javassist.NotFoundException;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class TestHazelcastSync {
    private static final Logger logger = LoggerFactory.getLogger(TestHazelcastSync.class);

    @Test
    public void testSlowJoin() throws Exception {
        try {
            CtClass clazz = ClassPool.getDefault().get("com.hazelcast.instance.Node");
            CtMethod m = clazz.getDeclaredMethod("setJoined");
            m.setBody("{" +
                    "logger.warning(\"Node joined delay\"); " +
                    "try {\n" +
                    "                        Thread.sleep(2000L);\n" +
                    "                    } catch (InterruptedException e) {\n" +
                    "                        e.printStackTrace();\n" +
                    "                    } " +
                    "joined.set(true);" +
                    "}");
            Class c = clazz.toClass();
        } catch (NotFoundException e) {
            e.printStackTrace();
        } catch (CannotCompileException e) {
            e.printStackTrace();
        }

        // create shared hazelcast config
        final Config config = new XmlConfigBuilder().build();
        String logging = "hazelcast.logging.type";
        System.setProperty(logging, "slf4j");
        System.setProperty("hazelcast.operation.generic.thread.count", "2");
        System.setProperty("hazelcast.operation.thread.count", "2");

        // get map config
        MapConfig mapConfig = config.getMapConfig("map1");

        // configure map store
        MapStoreConfig mapStoreConfig = new MapStoreConfig();
        mapStoreConfig.setEnabled(true);
        mapStoreConfig.setInitialLoadMode(MapStoreConfig.InitialLoadMode.LAZY);
        mapStoreConfig.setWriteDelaySeconds(10);
        mapStoreConfig.setWriteBatchSize(1);
        mapStoreConfig.setClassName("com.test.SlowMapStore");
        mapConfig.setMapStoreConfig(mapStoreConfig);

        // thread 1: start a single node
        Thread thread1 = new Thread(new Runnable() {

            @Override
            public void run() {
                HazelcastInstance hcInstance = Hazelcast.newHazelcastInstance(config);

                // try-finally to stop hazelcast instance
                try {
                    // log started
                    logger.info(Thread.currentThread().getName() + " started.");

                    // get map
                    // this will init the map (and call loadAllKeys())
                    IMap<String, String> map = hcInstance.getMap("map1");
                    int size = map.size();
                    logger.info("Map size = " + size);

                    // wait
                    Sleep.sleep(90000, true);
                } finally {
                    hcInstance.getLifecycleService().shutdown();
                }
            }
        }, "Thread 1");
        thread1.start();

        // wait 20s after starting first thread
        Sleep.sleep(30000, true);

        // thread 2: start a second node
        Thread thread2 = new Thread(new Runnable() {

            @Override
            public void run() {
                HazelcastInstance hcInstance = Hazelcast.newHazelcastInstance(config);

                // log joined
                logger.info(Thread.currentThread().getName() + " hazelcast instance joined.");

                // try-finally to stop hazelcast instance
                try {
                    // get map
                    IMap<String, String> map = hcInstance.getMap("map");
                    int size = map.size();
                    logger.info("Map size = " + size);

                    // wait before shutdown
                    Sleep.sleep(21000, true);
                } finally {
                    hcInstance.getLifecycleService().shutdown();
                }
                logger.info(Thread.currentThread().getName() + " done.");
            }
        }, "Thread 2");
        thread2.start();

        // join threads
        thread1.join();
        thread2.join();
    }
}

class Sleep {

    private static final Logger logger = LoggerFactory.getLogger(Sleep.class);

    public static void sleep(long ms, boolean log) {
        if (ms <= 0) {
            return;
        }
        try {
            Thread.sleep(ms);
            if (log) {
                logger.info("Slept " + (ms / 1000) + "s.");
            }
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }

}
```

```java
package com.test;

import com.hazelcast.core.MapStore;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeSet;
import java.util.concurrent.ConcurrentHashMap;

public class SlowMapStore implements MapStore<String, String> {

    private static final Logger logger = LoggerFactory.getLogger(SlowMapStore.class);

    // ----------------------------------------------------------------- config

    private final int sleepMsLoad;

    private final int sleepMsLoadAll;

    private final int sleepMsLoadAllPerKey;

    private final int sleepMsLoadAllKeys;

    // ------------------------------------------------------------------ state

    private final ConcurrentHashMap<String, String> store = new ConcurrentHashMap<String, String>();

    // ----------------------------------------------------------- construction

    public SlowMapStore() {
        this.sleepMsLoad = 20;
        this.sleepMsLoadAll = 50;
        this.sleepMsLoadAllPerKey = 10;
        this.sleepMsLoadAllKeys = 8000;
        preload(100);
        logger.info(String.format("JobsMapStore [jobs.map.store.s3.bucket] bucket name: %s, store name %s", "jobs-map-store-s3-bucket-dev-5", "jobs.map.store.s3.bucket"));
        Sleep.sleep(8000, true);
        logger.info("JobsMapStore [" + "jobs-map-store-s3-bucket-dev-5" + "] bucket: " + "jobs.map.store.s3.bucket");
    }

    public void preload(int size) {
        for (int i = 0; i < size; i++) {
            store.put("k" + i, "v" + i);
        }
    }

    // ---------------------------------------------------------------- getters

    // ----------------------------------------------------- MapStore interface

    @Override
    public String load(String key) {
        logger.info("load(" + key + ") called.");
        Sleep.sleep(sleepMsLoad, false);
        return store.get(key);
    }

    @Override
    public Map<String, String> loadAll(Collection<String> keys) {
        List<String> keysList = new ArrayList<String>(keys);
        Collections.sort(keysList);
        logger.info("loadAll(" + keysList + ") called.");
        Sleep.sleep(sleepMsLoadAll, false);
        Map<String, String> result = new HashMap<String, String>();
        for (String key : keys) {
            Sleep.sleep(sleepMsLoadAllPerKey, false);
            String value = store.get(key);
            if (value != null) {
                result.put(key, value);
            }
        }
        return result;
    }

    @Override
    public Set<String> loadAllKeys() {
        logger.info("loadAllKeys() called.");
        Sleep.sleep(sleepMsLoadAllKeys, true);
        Set<String> result = new HashSet<String>(store.keySet());
        List<String> resultList = new ArrayList<String>(result);
        Collections.sort(resultList);
        logger.info("loadAllKeys result: size = " + result.size() + ", keys = " + resultList + ".");
        return result;
    }

    @Override
    public void store(String key, String value) {
        logger.info("store(" + key + ") called.");
        store.put(key, value);
    }

    @Override
    public void storeAll(Map<String, String> map) {
        TreeSet<String> setSorted = new TreeSet<String>(map.keySet());
        logger.info("storeAll(" + setSorted + ") called.");
        store.putAll(map);
    }

    @Override
    public void delete(String key) {
        logger.info("delete(" + key + ") called.");
        store.remove(key);
    }

    @Override
    public void deleteAll(Collection<String> keys) {
        List<String> keysList = new ArrayList<String>(keys);
        Collections.sort(keysList);
        logger.info("deleteAll(" + keysList + ") called.");
        for (String key : keys) {
            store.remove(key);
        }
    }

}
```

This test leads to infinite sync for the 3.3.3:
```
2015-01-29 17:10:34,131|WARN |hz._hzInstance_2_dev.partition-operation.thread-1|c.h.p.InternalPartitionService|[192.168.3.65]:5702 [dev] [3.3.3] Sync replica target is null, no need to sync -> partition: 257, replica: 1
2015-01-29 17:10:34,132|DEBUG|hz._hzInstance_2_dev.partition-operation.thread-1|c.h.p.i.CheckReplicaVersion|[192.168.3.65]:5702 [dev] [3.3.3] Partition: 263 version is not matching to version of the owner -> 0 -vs- 1
2015-01-29 17:12:04,691|INFO |Thread 1|c.h.p.InternalPartitionService|[192.168.3.65]:5701 [dev] [3.3.3] Waiting for active migration tasks: 1
```
For the 3.4 it looks fine.

**Comment by mdogan** [comment]
Partition migration/replication mechanism has been changed/redesigned after 3.7+. Please try latest version and feel free to open a new issue if problem still exists.
